src/Features/CSharp/Portable/SignatureHelp/InvocationExpressionSignatureHelpProvider.cs

```diff
@@ -66,72 +66,99 @@ protected override async Task<SignatureHelpItems> GetItemsWorkerAsync(Document d
+            var anonymousTypeDisplayService = document.GetLanguageService<IAnonymousTypeDisplayService>();
+            var documentationCommentFormattingService = document.GetLanguageService<IDocumentationCommentFormattingService>();
+            IList<SignatureHelpItem> items;
+            int? selectedItem;
+            if (isDelegateType)
+            {
+                items = methods.SelectAsArray(m =>
+                    CreateItem(
+                        m, semanticModel, position,
+                        symbolDisplayService, anonymousTypeDisplayService,
+                        isVariadic: false,
+                        documentationFactory: null,
+                        prefixParts: GetDelegateInvokePreambleParts(m, semanticModel, position),
+                        separatorParts: GetSeparatorParts(),
+                        suffixParts: GetDelegateInvokePostambleParts(),
+                        parameters: GetDelegateInvokeParameters(m, semanticModel, position, documentationCommentFormattingService, cancellationToken)));
+                selectedItem = 0;
             }
             else
             {
-                return null;
+                (items, selectedItem) = GetMethodGroupItemsAndSelection(methods, currentSymbol,
+                    invocationExpression, semanticModel, symbolDisplayService,
+                    anonymousTypeDisplayService, documentationCommentFormattingService, cancellationToken);
             }
-        }
 
-        public override SignatureHelpState GetCurrentArgumentState(SyntaxNode root, int position, ISyntaxFactsService syntaxFacts, TextSpan currentSpan, CancellationToken cancellationToken)
-        {
-            if (TryGetInvocationExpression(
-                    root,
-                    position,
-                    syntaxFacts,
-                    SignatureHelpTriggerReason.InvokeSignatureHelpCommand,
-                    cancellationToken,
-                    out var expression) &&
-                currentSpan.Start == SignatureHelpUtilities.GetSignatureHelpSpan(expression.ArgumentList).Start)
+            var textSpan = SignatureHelpUtilities.GetSignatureHelpSpan(invocationExpression.ArgumentList);
+            if (currentSymbol is null || parameterIndex < 0)
             {
-                return SignatureHelpUtilities.GetSignatureHelpState(expression.ArgumentList, position);
+                var argumentIndex = GetArgumentIndex(arguments, position);
+                return new SignatureHelpItems(items, textSpan, argumentIndex < 0 ? 0 : argumentIndex, arguments.Count, argumentName: null, selectedItem);
             }
 
-            return null;
+            var methodSymbol = (IMethodSymbol)currentSymbol;
+            var parameters = methodSymbol.Parameters;
+            var name = parameters.Length == 0 ? null : parameters[parameterIndex].Name;
+            return new SignatureHelpItems(items, textSpan, parameterIndex, methodSymbol.Parameters.Length, name, selectedItem);
```

Comment:
ok.  i've now seen what looks like the same code *twice*.  Once for constructors, once for delegates.  I am *not* a fan.  Is it not possible to share more between the two?  I'll be especially bummed if i keep going and i see this for more sig-help cases below.